block-storage/src/main/scala/coop/rchain/blockstorage/util/DoublyLinkedDagOperations.scala

```diff
@@ -60,7 +62,12 @@ object DoublyLinkedDagOperations {
 
   // If the element doesn't exist in the dag, the dag is returned as is
   @SuppressWarnings(Array("org.wartremover.warts.Throw")) // TODO remove throw
-  def remove[A](dag: DoublyLinkedDag[A], element: A): DoublyLinkedDag[A] = {
+  def remove[A](
+      dag: DoublyLinkedDag[A],
+      element: A
+  ): (DoublyLinkedDag[A], mutable.Set[A], mutable.Set[A]) = {
+    val childrenAffected                           = mutable.Set.empty[A]
+    val childrenRemoved                            = mutable.Set.empty[A]
```

Comment:
Better to use immutable Set. Mutable structure if contained inside function can be justified but returning mutable object is calling for trouble.

Reply:
Is there an elegant way to use immutable Set taking into account that we disabled var? Using e.g. Ref just to update immutable structure looks excessive.  For now I'm converting mutable set to immutable on return.

Reply:
It's easy to convert this code to use immutable structure. This additional state should be also part of accumulator used when folding over children.
```scala
val initAcc = (childToParentAdjacencyList, Set.empty[A], Set.empty[A], Set.empty[A])
```
Maybe even helper case class would be useful, to be more clear about this temporary state.